pkg/multicluster/watcher.go

```diff
@@ -0,0 +1,193 @@
+
+const (
+	LocalCluster = ""
+)
+
+type ClusterWatcher interface {
+	// Run starts a watch for KubeConfig secrets on the cluster managed by the local manager.Manager.
+	// Note that Run will call Start on the local manager and run all registered ClusterHandlers.
+	Run(local manager.Manager) error
+	// RegisterClusterHandler adds a ClusterHandler to the ClusterWatcher.
+	RegisterClusterHandler(handler ClusterHandler)
+	// Stop cancels the context on every manager.Manager.
+	Stop()
+}
+
+type clusterWatcher struct {
+	ctx      context.Context
+	mutex    sync.RWMutex
+	handlers *handlerList
+	cancels  *cancelSet
+}
+
+var _ ClusterWatcher = &clusterWatcher{}
+
+func NewClusterWatcher(ctx context.Context) *clusterWatcher {
+	return &clusterWatcher{
+		ctx:      ctx,
+		mutex:    sync.RWMutex{},
+		handlers: newHandlerList(),
+		cancels:  newCancelSet(),
+	}
+}
+
+func (c *clusterWatcher) Run(local manager.Manager) error {
+	c.startManager(LocalCluster, local)
+	loop := controller.NewSecretReconcileLoop("cluster watcher", local)
+	return loop.RunSecretReconciler(c.ctx, c, kubeconfig.SecretPredicate{})
+}
+
+func (c *clusterWatcher) ReconcileSecret(obj *v1.Secret) (reconcile.Result, error) {
+	clusterName, cfg, err := kubeconfig.SecretToConfig(obj)
+	if err != nil {
+		return reconcile.Result{}, eris.Wrap(err, "failed to extract kubeconfig from secret")
+	}
+
+	// If the cluster already has a manager, remove the existing instance and start again.
+	if _, err := c.cancels.get(clusterName); err == nil {
+		c.removeCluster(clusterName)
+	}
+
+	mgr, err := manager.New(cfg.RestConfig, manager.Options{
+		// TODO these should be configurable, disable for now
+		MetricsBindAddress:     "0",
+		HealthProbeBindAddress: "0",
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	c.startManager(clusterName, mgr)
+
+	return reconcile.Result{}, nil
+}
+
+func (c *clusterWatcher) startManager(clusterName string, mgr manager.Manager) {
+	ctx, cancel := context.WithCancel(
+		contextutils.WithLoggerValues(context.WithValue(c.ctx, "cluster", clusterName), zap.String("cluster", clusterName)))
+	go func() {
+		err := mgr.Start(ctx.Done())
+		if err != nil {
+			contextutils.LoggerFrom(ctx).DPanicw("manager start failed for cluster %v", clusterName)
+		}
+	}()
+
+	c.cancels.set(clusterName, cancel)
+	c.handlers.HandleAddCluster(ctx, clusterName, mgr)
+}
+
+func (c *clusterWatcher) ReconcileSecretDeletion(req reconcile.Request) {
+	// TODO update to namespace.name
+	c.removeCluster(req.Name)
+}
+
+func (c *clusterWatcher) removeCluster(clusterName string) {
+	// TODO joekelley update cancels.delete to also call cancel
+	cancel, err := c.cancels.get(clusterName)
+	if err != nil {
+		contextutils.LoggerFrom(c.ctx).Debugw("reconciled delete on cluster secret for uninitialized cluster %v", clusterName)
+		return
+	}
+
+	cancel()
+	c.cancels.delete(clusterName)
+}
+
+func (c *clusterWatcher) RegisterClusterHandler(handler ClusterHandler) {
+	c.handlers.add(handler)
```

Comment:
Do we want to retroactively add existing clusters?

Reply:
Reading this as "call the new ClusterHandler's AddCluster function for all clusters already discovered by the watcher", disregard if I misunderstood.

good question, any thoughts on this @ilackarms ? Currently the watcher is not responsible for maintaining the active set of managers, just their cancel functions, so we'd need a refactor to support this behavior.

A multicluster client will be introduced in a follow-up PR. The plan there is to have the client register with the watch as a ClusterHandler and maintain the set of managers created by the ClusterWatcher. If we decide we want the feature proposed here, it may make sense for the watcher to be responsible for storing active managers and adding a function to expose them.

Reply:
So I think it depends whether we think handlers should have all previously added clusters, added upon startup. If so I would definitely lean towards having it here. If we push it down a level, then that level becomes the new root, and it sounds like we want the multicluster client to be just like any other handler.

To solve this, we could just change the cancelSet, to a set containing a struct such as:
```
type ContextualManager struct {
    ctx context.Context
    cancel context.CancelFunc
    mgr manager.manager
}
```

This would also allow flexibility to add things like the errSignaler I mentioned above.

Reply:
I made these changes. It necessitated the addition of a ClusterHandler per reconciler.

Now, when adding a reconcile loop, we first register the loop itself with the ClusterWatcher. The loop is a ClusterHandler whose AddCluster function creates a reconcile.Loop for every cluster. Its AddReconciler function now creates a new object called a `reconcilerRunner` which also implements the ClusterHandler interface, and registers it to the ClusterWatcher. This allows us to add reconcilers to a multicluster reconcile loop after the ClusterWatcher has been started.

Reply:
cleaned up https://github.com/solo-io/skv2/pull/4/commits/b6f7f2b2bccc31f7ce22c03715f596f008844f81